In [292]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from os import listdir

import math

import warnings
warnings.filterwarnings('ignore')

## Uploading the data

In [293]:
def upload_data(data):
    spontan_list= [] 
    sync_list= []
    alone_list = []
    for participant in listdir(complete_data):
        for file in listdir(complete_data+participant+'/'):
            if file.endswith('Spontan.csv'):
                spontan=pd.read_csv(complete_data+participant+'/'+file)
            elif file.endswith('Sync.csv'):
                sync=pd.read_csv(complete_data+participant+'/'+file)
            elif file.endswith('Alone.csv'):
                alone=pd.read_csv(complete_data+participant+'/'+file)
        spontan_list.append(spontan)
        sync_list.append(sync)
        alone_list.append(alone)
    return spontan_list,sync_list,alone_list

In [294]:
#train set
complete_data=(r'C:\Users\meirk\Desktop\final_project_files\finalProjectFiles\extraFiles\Training/')
spontan_list, sync_list ,alone_list = upload_data(complete_data)
print ("number of files is: ",len(spontan_list),',',len(sync_list),',',len(alone_list),
       "in spontan, sync and alone respectively")
#shapes of dfs
train_list = [spontan_list,sync_list,alone_list]
for list_ in train_list:
    for file in list_:
        print(file.shape)

number of files is:  9 , 9 , 9 in spontan, sync and alone respectively
(9054, 22)
(8998, 22)
(9050, 22)
(8937, 22)
(8826, 22)
(9054, 22)
(9054, 22)
(8773, 22)
(4375, 22)
(9052, 22)
(9054, 22)
(9039, 22)
(9050, 22)
(8395, 22)
(9046, 22)
(9054, 22)
(8829, 22)
(4500, 22)
(4526, 22)
(4575, 22)
(4482, 22)
(4527, 22)
(4931, 22)
(4526, 22)
(4468, 22)
(4537, 22)
(2250, 22)


In [295]:
#validation set
complete_data=(r'C:\Users\meirk\Desktop\final_project_files\finalProjectFiles\extraFiles\Validation/')
spontan_list_val, sync_list_val ,alone_list_val = upload_data(complete_data)
print ("number of files in is: ",len(spontan_list),',',len(sync_list),',',len(alone_list),
       "in spontan, sync and alone respectively")
#shapes of dfs
validation_list = [spontan_list_val, sync_list_val ,alone_list_val]
for list_ in validation_list:
    for file in list_:
        print(file.shape)

number of files in is:  9 , 9 , 9 in spontan, sync and alone respectively
(8969, 22)
(8986, 22)
(4500, 22)
(8846, 22)
(8965, 22)
(4446, 22)
(4700, 22)
(4614, 22)
(2250, 22)


We uploaded the data (both training and validation sets) into 3 lists, one for each scenario. Each list contains 9 files, a file for each participant.

In [296]:
#HandRight
right_hand = pd.read_csv(r'C:\Users\meirk\Desktop\final_project_files\finalProjectFiles\extraFiles\HandRight.csv')
right_hand.head()

Time   Frame ID  Hand Type   # hands   Position X   Position Y  \
0  128.2883      15478      right         1     15.67167     226.2064   
1  128.3048      15480      right         1     16.24119     222.9389   
2  128.3218      15482      right         1     16.72461     220.9709   
3  128.3384      15484      right         1     17.09016     219.7124   
4  128.3551      15486      right         1     17.46017     219.3652   

    Position Z   Velocity X   Velocity Y   Velocity Z  ...       Yaw  \
0     20.26470     34.26665  -195.214700   -35.043020  ... -0.277800   
1     20.11781     36.16443  -180.093900    -4.052799  ... -0.277763   
2     19.94972     26.35298   -95.088190   -11.157620  ... -0.278792   
3     19.67679     19.46804   -61.755180   -21.303780  ... -0.279768   
4     19.13128     21.94964    -8.355943   -35.081790  ... -0.278629   

    Wrist Pos X   Wrist Pos Y   Wrist Pos Z   Elbow pos X   Elbow Pos Y  \
0      46.88831      207.3548      86.23589      203.8337      49.33507   
1      47.54861      205.6568      86.42250      203.9553      47.88272   
2      48.15030      204.5182      86.35149      204.6231      45.35921   
3      48.64175      203.9989      86.16056      204.9939      44.18692   
4      48.93135      203.3777      85.59017      204.5952      42.68003   

    Elbow Pos Z   Grab Strenth   Grab Angle   Pinch Strength  
0      237.4553            0.0     0.206389              0.0  
1      238.4482            0.0     0.243050              0.0  
2      236.8544            0.0     0.281134              0.0  
3      236.0927            0.0     0.277691              0.0  
4      235.2896            0.0     0.283008              0.0  

[5 rows x 22 columns]

In [297]:
right_hand.shape

(4812, 22)

## preprocessing

###### Search for NaN values

In [298]:
nan_sum = 0
#train set
print("train set")
for list_ in train_list:
    for file in list_:
        nan_sum += file.isna().sum().sum()
print("number of NaN values in train set is: ",nan_sum)
#validation set
print("validation set")
validation_list = [spontan_list_val, sync_list_val ,alone_list_val]
for list_ in validation_list:
    for file in list_:
        nan_sum = file.isna().sum().sum()
print("number of NaN values in validation test is: ",nan_sum)
#right hand
print("right hand set")
print("number of NaN values in right hand set is: ",right_hand.isna().sum().sum())

train set
number of NaN values in train set is:  0
validation set
number of NaN values in validation test is:  0
right hand set
number of NaN values in right hand set is:  2304


right_hand  dataFrame has 2304 NaN values.

In [299]:
right_hand.isna().sum()

Time                 0
 Frame ID            0
 Hand Type           0
 # hands             0
 Position X          0
 Position Y          0
 Position Z          0
 Velocity X          0
 Velocity Y          0
 Velocity Z          0
 Pitch               0
 Roll                0
 Yaw                 0
 Wrist Pos X       384
 Wrist Pos Y       384
 Wrist Pos Z       384
 Elbow pos X       384
 Elbow Pos Y       384
 Elbow Pos Z       384
 Grab Strenth        0
 Grab Angle          0
 Pinch Strength      0
dtype: int64

As we already saw, right hand df has 4812 values. Without NaN number it contains:

In [300]:
print(len(right_hand.index) - len(right_hand.dropna().index))

384


384 frames contains NaN values, lets drop them.

In [301]:
right_hand = right_hand.dropna()
len(right_hand.index)

4428

###### converting 'Hand Type' to ones - right and zeros - left

In [302]:
def convert_to_1and0 (column):
    for index in range(len(column)):
        if column.loc[index] == 'right':
            column.loc[index] = 1
        else:
            column.loc[index] = 0    

In [303]:
#train set
for list_ in train_list:
    for df in list_:
        df[' Hand Type'] = pd.to_numeric(df[' Hand Type'], errors= 'coerce')
        convert_to_1and0(df[' Hand Type'])
#validation set
for list_ in validation_list:
    for df in list_:
        df[' Hand Type'] = pd.to_numeric(df[' Hand Type'], errors= 'coerce')
        convert_to_1and0(df[' Hand Type'])
#right hand set
right_hand[' Hand Type'] = pd.to_numeric(right_hand[' Hand Type'], errors= 'coerce')
convert_to_1and0(right_hand[' Hand Type'])

Checking on the train set if the method above worked as expected

In [304]:
i = 1
#train set
for list_ in train_list:
    for df in list_:
        print("df number ",i)
        print(df[' Hand Type'].value_counts())
        i = i + 1

df number  1
0.0    9054
Name:  Hand Type, dtype: int64
df number  2
0.0    8998
Name:  Hand Type, dtype: int64
df number  3
0.0    9050
Name:  Hand Type, dtype: int64
df number  4
0.0    8937
Name:  Hand Type, dtype: int64
df number  5
0.0    8826
Name:  Hand Type, dtype: int64
df number  6
0.0    9054
Name:  Hand Type, dtype: int64
df number  7
0.0    9054
Name:  Hand Type, dtype: int64
df number  8
0.0    8773
Name:  Hand Type, dtype: int64
df number  9
0.0    4375
Name:  Hand Type, dtype: int64
df number  10
0.0    9052
Name:  Hand Type, dtype: int64
df number  11
0.0    9054
Name:  Hand Type, dtype: int64
df number  12
0.0    9039
Name:  Hand Type, dtype: int64
df number  13
0.0    9050
Name:  Hand Type, dtype: int64
df number  14
0.0    8395
Name:  Hand Type, dtype: int64
df number  15
0.0    9046
Name:  Hand Type, dtype: int64
df number  16
0.0    9054
Name:  Hand Type, dtype: int64
df number  17
0.0    8829
Name:  Hand Type, dtype: int64
df number  18
0.0    4500
Name:  Hand Ty

It worked well.

In [305]:
spontan_list[0].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9054 entries, 0 to 9053
Data columns (total 22 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Time             9054 non-null   float64
 1    Frame ID        9054 non-null   int64  
 2    Hand Type       9054 non-null   float64
 3    # hands         9054 non-null   int64  
 4    Position X      9054 non-null   float64
 5    Position Y      9054 non-null   float64
 6    Position Z      9054 non-null   float64
 7    Velocity X      9054 non-null   float64
 8    Velocity Y      9054 non-null   float64
 9    Velocity Z      9054 non-null   float64
 10   Pitch           9054 non-null   float64
 11   Roll            9054 non-null   float64
 12   Yaw             9054 non-null   float64
 13   Wrist Pos X     9054 non-null   float64
 14   Wrist Pos Y     9054 non-null   float64
 15   Wrist Pos Z     9054 non-null   float64
 16   Elbow pos X     9054 non-null   float64
 17   Elbow Pos Y  

Hand Type was changed to float64 instead of Object.

###### Removing first 7 seconds

In [306]:
def remove_7_seconds(df):
    time_to_drop = df['Time'][0]
    time_to_drop = time_to_drop + 7
    print("time to drop for df is ",time_to_drop)
    while df['Time'][0] < time_to_drop:
        df = df.drop(0, axis = 0)
       # frame_num += 1
    return df

In [307]:
#train setdf = df.drop(df[df.score < 50].index)
for list_ in train_list:
    for df in list_:
        time_to_drop = df['Time'][0]
        time_to_drop = time_to_drop + 7
        print("time to drop for df is ",time_to_drop)
        df_ = df.drop(df[df['Time']<time_to_drop].index, inplace=True)
    
     #   frame_num = 0
      #  while df['Time'][frame_num] < time_to_drop:
       #     df = df.drop(frame_num, axis = 0)
        #    frame_num += 1
#validation set
for list_ in validation_list:
    for df in list_:
        time_to_drop = df['Time'][0]
        time_to_drop = time_to_drop + 7
        print("time to drop for df is ",time_to_drop)
        df_ = df.drop(df[df['Time']<time_to_drop].index, inplace= True)
     #   frame_num = 0
      #  while df['Time'][frame_num] < time_to_drop:
       #     df = df.drop(frame_num, axis = 0)
        #    frame_num += 1

time to drop for df is  342.9984
time to drop for df is  392.3023
time to drop for df is  533.4969
time to drop for df is  382.6675
time to drop for df is  568.7278
time to drop for df is  386.7184
time to drop for df is  499.7171
time to drop for df is  182.5415
time to drop for df is  605.6176
time to drop for df is  434.231
time to drop for df is  613.8324
time to drop for df is  706.221
time to drop for df is  471.5643
time to drop for df is  654.2784
time to drop for df is  480.3199
time to drop for df is  701.4501
time to drop for df is  298.0203
time to drop for df is  788.3885
time to drop for df is  525.8087
time to drop for df is  503.0675
time to drop for df is  619.9247
time to drop for df is  295.7583
time to drop for df is  464.7846
time to drop for df is  568.4594
time to drop for df is  587.1392
time to drop for df is  393.9574
time to drop for df is  696.6486
time to drop for df is  779.2799
time to drop for df is  63.73743
time to drop for df is  823.9093
time to drop

In [308]:
right_hand_ = right_hand.copy()
time_to_drop = right_hand_['Time'][0]
time_to_drop = time_to_drop + 7
print("time to drop for hand right df is ",time_to_drop)
right_hand_ = right_hand_.drop(right_hand_[right_hand_['Time']<time_to_drop].index, inplace= True)
#print("check: 'Time' of first row after dropping: ",right_hand_['Time'][0])

time to drop for hand right df is  135.2883
